File: problems/9.recursion/ts/quick-select.ts
function swap(arr: number[], i: number, j: number) {
    let temp = arr[j];
    arr[j] = arr[i];
    arr[i] = temp;
}

function partition(arr: number[], left: number, right: number) {
    const pivotValue = arr[right];
    let partitionIndex = left - 1;

    for (let i = left; i < right; i++) {
        if (pivotValue > arr[i]) {
            ++partitionIndex;
            swap(arr, partitionIndex, i);
        }
    }

    swap(arr, partitionIndex + 1, right);
    return partitionIndex + 1;
}

export function quickSelect(
    arr: number[],
    left: number,
    right: number,
    indexToFind: number
): number {
    if (left >= right) {
        return arr[indexToFind];
    }

    const partitionIndex = partition(arr, left, right);

    if (partitionIndex === indexToFind) {
        return arr[indexToFind];
    } else if (partitionIndex > indexToFind) {
        return quickSelect(arr, left, partitionIndex - 1, indexToFind);
    } else {
        return quickSelect(arr, partitionIndex + 1, right, indexToFind);
    }
}

export function returnKthLargestElement(arr: number[], k: number) {
    return quickSelect(arr, 0, arr.length - 1, arr.length - k);
}

File: golang/6.sort/quick_select.go
package main

func quickSelect(arr []int, left, right, indexToFind int) int {
	if left >= right {
		return arr[indexToFind]
	}

	partitionIndex := partition(arr, left, right)
	if indexToFind == partitionIndex {
		return arr[partitionIndex]
	} else if indexToFind < partitionIndex {
		return quickSelect(arr, left, partitionIndex-1, indexToFind)
	} else {
		return quickSelect(arr, partitionIndex+1, right, indexToFind)
	}
}

func returnKthLargestElement(arr []int, k int) int {
	return quickSelect(arr, 0, len(arr)-1, len(arr)-k)
}


